plugins/audit-log-node/src/DefaultAuditLogger.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  AuthService,
+  HttpAuthService,
+  LoggerService,
+} from '@backstage/backend-plugin-api';
+import { ErrorLike } from '@backstage/errors';
+
+import { Request } from 'express';
+
+import {
+  AuditErrorLogOptions,
+  AuditLogDetailsOptions,
+  AuditLogger,
+  AuditLoggerOptions,
+  AuditLogOptions,
+} from './types';
+
+export class DefaultAuditLogger implements AuditLogger {
+  private readonly logger: LoggerService;
+  private readonly authService: AuthService;
+  private readonly httpAuthService: HttpAuthService;
+
+  constructor(options: AuditLoggerOptions) {
+    this.logger = options.logger;
+    this.authService = options.authService;
+    this.httpAuthService = options.httpAuthService;
+  }
+
+  async getActorId(request?: Request): Promise<string | undefined> {
+    if (!(request && this.httpAuthService && this.authService)) {
+      return undefined;
+    }
+    try {
+      const credentials = await this.httpAuthService.credentials(request);
+      const userEntityRef = this.authService.isPrincipal(credentials, 'user')
+        ? credentials.principal.userEntityRef
+        : undefined;
+
+      const serviceEntityRef = this.authService.isPrincipal(
+        credentials,
+        'service',
+      )
+        ? credentials.principal.subject
+        : undefined;
+
+      return userEntityRef ?? serviceEntityRef;
+    } catch {
+      return undefined;
+    }
+  }
+  async createAuditLogDetails(options: AuditLogDetailsOptions) {
+    const { eventName, stage, metadata, actor_id, request, response, status } =
+      options;
+
+    const actorId = actor_id || (await this.getActorId(request)) || null;
+
+    // Secrets in the body field should be redacted by the user before passing in the request object
+    const auditRequest = request
+      ? {
+          method: request.method,
+          url: request.originalUrl,
+          params: request.params,
+          query: request.query,
+          body: request.body,
+        }
+      : undefined;
+
+    const auditLogCommonDetails = {
```

Comment:
For the RBAC backend plugin, some parts of the changes come from the policy file or the plugin itself, so we do not always have the IP address, hostname, or user agent. Therefore, I propose changing this code to:
  
```
    let actor: ActorDetails = { actorId };
    if (request) {
      actor.ip = request.ip,
      actor.hostname = request.hostname,
      actor.userAgent = request.get('user-agent')
    }
    
    const auditLogCommonDetails = {
      actor,
      meta: metadata || {},
      request: auditRequest,
      isAuditLog: true as const,
      response,
      eventName,
      stage,
    };
  ```
  
Then, this code will simply skip these fields when the request is absent, instead of printing undefined values.

Reply:
Hmm. I guess it would be nice to do the same with request and response:

```
    let actor: ActorDetails = { actorId };
    if (request) {
      actor.ip = request.ip;
      actor.hostname = request.hostname;
      actor.userAgent = request.get('user-agent');
    }

    const auditLogCommonDetails: AuditLogDetails = {
      actor,
      meta: metadata || {},
      isAuditLog: true as const,
      eventName,
      stage,
      status: 'succeeded',
    };

    if (auditRequest) {
      auditLogCommonDetails.request = auditRequest;
    }
    if (response) {
      auditLogCommonDetails.response = response;
    }

    if (status === 'failed') {
      const errs = options.errors as ErrorLike[];
      return {
        ...auditLogCommonDetails,
        status,
        errors: errs.map(err => {
          return {
            name: err.name,
            message: err.message,
            stack: err.stack,
          };
        }),
      };
    }

    return {
      ...auditLogCommonDetails,
    };
   ```

Reply:
I mainly kept the `undefined` printing since keys with `undefined` values don't actually get printed. I can make those changes to make it more verbose as well.

Reply:
@kim-tsao, would customers prefer a field to be null or to not exist at all (undefined)? I believe they might prefer undefined, as it would take up less storage.

Reply:
@AndrienkoAleksandr can you clarify this scenario?

> For the RBAC backend plugin, some parts of the changes come from the policy file or the plugin itself,

Is this the case where a cluster admin makes a change to the config file?  If so, they should enable audit logging on the cluster to capture these events.  Generating audit logs without any of the required actor identifiers would not be useful at all.  Can we can suppress audit log generation as long as we are certain this is the _only_ scenario where actor identifiers cannot be obtained?

We will need to document that Customers need to turn on audit logging in the clusters if they are making changes there.

Reply:
Yes, @AndrienkoAleksandr is referencing whenever a cluster admin makes changes to the csv file in the cluster, those changes are picked up by the pod, and then our plugin pulls them in.

Reply:
In my case for the scaffolder-backend, many of the actions are being done asynchronously by the backend, so I just manually set an `actorId` of `scaffolder-backend`.

Reply:
@Zaperex I thought about having a generic ID but I was a bit concerned that we could be generating too many events.  Maybe this won't be a problem since configmap modifications shouldn't be too frequent? I guess it doesn't hurt to generate additional records with the generic ID as long as it's not too excessive.   We can consider generic ID events as "backend triggered" which customers can then use to investigate further with their cluster audit logs

Reply:
@kim-tsao I'm also going to point out that a good portion of the `scaffolder-backend` audit logging has a generic actorID since I'm audit logging each scaffolder task (and all it's steps/iterations), but it only logs when tasks are being executed so it will not always be on. Will there be any concerns there?

However, the amount of generic actorID events being generated by the `catalog-backend` will definitely be alot as the catalog will be constantly asynchronously making updates to the entities without any user interactions, so we might need to discuss what details we want to audit log or perhaps log some of the audit logs to a `debug` log level or something? Ex: this catalog entity was updated from a scheduled refresh. What do you think?

Reply:
@Zaperex 

> I'm also going to point out that a good portion of the scaffolder-backend audit logging has a generic actorID since I'm audit logging each scaffolder task (and all it's steps/iterations), but it only logs when tasks are being executed so it will not always be on. Will there be any concerns there?

Can these generic audit logs be tied back to the software template (id perhaps?) execution that was initiated by the user?  

> However, the amount of generic actorID events being generated by the catalog-backend will definitely be alot as the catalog will be constantly asynchronously making updates to the entities without any user interactions, so we might need to discuss what details we want to audit log or perhaps log some of the audit logs to a debug log level or something? Ex: this catalog entity was updated from a scheduled refresh. What do you think?

This should be ok as long as it was a user interaction that caused the async update (I think we talked about having a unique identifier tie these two events together at one point in time).  If it's a service sync refresh, we can consider that out of scope.

Reply:
@kim-tsao 
> Can these generic audit logs be tied back to the software template (id perhaps?) execution that was initiated by the user?

Yep they can, most of the scaffolder related audit logs have a `taskId` metadata field for the corresponding task. There would also be an audit log specifying which user created a task with that `taskId`. It's also stored in the scaffolder database alongside the tasks. I was mainly asking if there would be concerns with the volume of events being audit logged with the scaffolder and catalog since you said `I thought about having a generic ID but I was a bit concerned that we could be generating too many events`

> This should be ok as long as it was a user interaction that caused the async update (I think we talked about having a unique identifier tie these two events together at one point in time). If it's a service sync refresh, we can consider that out of scope.

Pretty much all the changes made to the catalog are caused by these catalog service sync refresh (a refresh essentially is an entity getting processed/re-processed), so it's quite difficult to differentiate between ones triggered by explicitly caused by user interactions and ones caused by the backend since everything is added to the same processing queue. 

Ex: Suppose the `catalog-info.yaml` that the location entity points to changed, then the corresponding entity will be updated the next time it gets queued up to be processed. However, a user manually triggering a refresh would cause the same process to occur with the only difference being that the user put the entity/location higher up in the queue.